internal/worker/run.go

```diff
@@ -191,7 +211,89 @@ func (w *Worker) runStep(ctx context.Context, log logr.Logger, sandbox cri.CRI,
 	log.Info(aurora.Green("Step completed").String())
 }
 
-func (w *Worker) runTaskLogic(ctx context.Context, syslog, log logr.Logger, slot int, claim *tcqueue.TaskClaim, wr io.Writer) error {
+func (w *Worker) uploadArtifact(
+	ctx context.Context, log logr.Logger, claim *tcqueue.TaskClaim,
+	rootContainer cri.Container, artifact *config.Artifact,
+	deps []<-chan error, ch chan<- error,
+) {
+	var err error
+	defer func() {
+		ch <- err
+		close(ch)
+	}()
+
+	// Wait for all dependencies.
+	for _, ch := range deps {
+		err = <-ch
+		if err != nil {
+			return
+		}
+	}
+
+	if artifact.Type != config.FileArtifact {
+		err = fmt.Errorf("only file artifacts are supported")
+		return
+	}
+
+	srcPath := artifact.Path
+	if !path.IsAbs(srcPath) {
+		srcPath = path.Join("/workspace", artifact.Path)
+	}
+
+	var r io.ReadCloser
+	r, info, err := rootContainer.ReadFiles(ctx, srcPath)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if err := r.Close(); err != nil {
+			log.Error(err, "failed to close artifact file")
+		}
+	}()
+
+	if info.IsDir() {
+		err = fmt.Errorf("directory artifacts are not supported")
+		return
+	}
+
+	expiresAt := time.Time(claim.Task.Expires)
+	size := info.Size()
+
+	tr := tar.NewReader(r)
+	header, err := tr.Next()
+	if err != nil {
+		return
+	}
+
+	log.Info("uploading artifact",
+		"name", artifact.Name,
+		"path", artifact.Path,
+		"size", size,
+	)
+
+	if header.Size != info.Size() {
+		err = fmt.Errorf("file is wrong size")
+		return
+	}
+
+	credentials := taskCredentials(&claim.Credentials)
+	queue := tcqueue.New(credentials, w.config.RootURL)
+	queue.Context = ctx
+	err = createS3Artifact(queue, claim, artifact.Name, artifact.ContentType, expiresAt, size, tr)
+	if err != nil {
+		return
+	}
+
+	_, err = tr.Next()
+	if err != io.EOF {
+		err = fmt.Errorf("unexpected tar files")
+		return
+	}
```

Comment:
artifacts must be single files, which are tar'ed, untar'ed, and submitted to TC? this feels weird? is this a TC-ism?

Reply:
Nope, just a Docker-ism.

Reply:
Amazing

-ly horrifying.